### PR TITLE
Make SamAuthProvider more performant

### PR DESCRIPTION
**Pull request by rtitle**
DataBiosphere/leonardo#189

Friends with https://github.com/broadinstitute/firecloud-develop/pull/970

1. Cache results of `SamAuthProvider.hasNotebookClusterPermission()` and `SamAuthProvider.canSeeAllClustersInProject()` because those methods are called very often by the proxy and "get/list clusters" endpoints.
   - `SamAuthProvider.hasProjectPermission()` is NOT cached because that is only called by "create cluster" which is less frequent.

2. Enforce a configurable timeout on all provider calls. The timeout setting is supported for ANY provider impl, defaults to 30 seconds, and is overridable by the provider if they wish.

3. Minor refactoring: put Sam-related provider classes in a `sam` subpackage, and other things I'll note inline

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: leonardo-example.conf [hunk 1]
```diff
@@ -60,9 +60,9 @@ clusterFiles {
 akka {
   loglevel = INFO
   http {
-    server.idle-timeout = 1 h
-    client.idle-timeout = 1 h
-    server.request-timeout = 60 s
+    server.idle-timeout = 1 hour
```
Comment (rtitle) [inline leonardo-example.conf]: OCD: changed duration configs to be in a consistent format

File: leonardo-example.conf [hunk 2]
```diff
@@ -110,8 +115,13 @@ auth {
 serviceAccounts {
   # To implement your own service account provisioning, subclass org.broadinstitute.dsde.workbench.leonardo.model.ServiceAccountProvider
   # and specify your concrete class here
-  providerClass = ""
-  config { # Do not remove config, even if empty
```
Comment (rtitle) [inline leonardo-example.conf]: I had previously renamed this key to `providerConfig` to be consistent with the other provider, but forgot to update the example conf.

File: src/main/resources/reference.conf [hunk 3]
```diff
@@ -50,23 +50,39 @@ proxy {
...
 # Authorization implementation config
-# TODO remove this when we remove the whitelist config
 auth {
-  # The other option is to provide a whitelist of users
-  whitelistProviderConfig = {
+  providerConfig {
+    # Amount of time Leo will wait for a provider response before timing out.
+    # This should be set less than akka.http.server.request-timeout.
+    # This option is supported for ANY provider implementation. If not specified, the default is 30 seconds.
+    providerTimeout = 30 seconds
```
Comment (rtitle) [inline src/main/resources/reference.conf]: Timeout is specified here and _not_ in firecloud-develop

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/DefaultServiceAccountProvider.scala [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import org.broadinstitute.dsde.workbench.leonardo.model.ServiceAccountProvider
+import org.broadinstitute.dsde.workbench.model.WorkbenchEmail
+import org.broadinstitute.dsde.workbench.model.google.GoogleProject
+
+import scala.concurrent.{ExecutionContext, Future}
+
+/**
+  * A ServiceAccountProvider which uses all defaults.
+  */
+class DefaultServiceAccountProvider(config: Config) extends ServiceAccountProvider(config) {
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/DefaultServiceAccountProvider.scala]: A trivial `ServiceAccountProvider` impl. This is referenced in the Leo example conf.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/LeoAuthProviderHelper.scala [hunk 5]
```diff
@@ -44,7 +52,7 @@ class LeoAuthProviderHelper(wrappedAuthProvider: LeoAuthProvider, authConfig: Co
     }
 
     // recover from failed futures AND catch thrown exceptions
-    try { future.recoverWith(exceptionHandler) } catch exceptionHandler
+    try { future.withTimeout(providerTimeout, s"LeoAuthProvider timed out after $providerTimeout").recoverWith(exceptionHandler) } catch exceptionHandler
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/LeoAuthProviderHelper.scala]: Note the timeout is applied before the recovery, so a `TimeoutException` is mapped to a `LeoException` which is handled by our app!

Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/LeoAuthProviderHelper.scala]: Also note: `withTimeout` comes from here: https://github.com/broadinstitute/workbench-libs/blob/develop/util/src/main/scala/org/broadinstitute/dsde/workbench/util/FutureSupport.scala#L51

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/sam/SamProvider.scala [hunk 6]
```diff
@@ -17,8 +17,8 @@ trait SamProvider {
   protected def getLeoServiceAccountAndKey: (WorkbenchEmail, File)
 
   protected lazy val samServer = config.as[String]("samServer")
-  protected lazy val cacheExpiryTime = toScalaDuration(Option(config.getDuration("samCacheExpiryTime")).getOrElse(Duration.ofMinutes(60)))
-  protected lazy val cacheMaxSize = Option(config.as[Int]("samCacheMaxSize")).getOrElse(1000)
+  protected lazy val cacheExpiryTime = config.getAs[FiniteDuration]("petTokenCacheExpiryTime").getOrElse(60 minutes)
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/sam/SamProvider.scala]: Ficus better

**Comment by rtitle** [comment]
jenkins retest

File: leonardo-example.conf (hunk 2 above)
Comment (helgridly) [reply leonardo-example.conf]: thank you for this change, it's been quietly making me go mad

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/LeoAuthProviderHelper.scala (hunk 5 above)
Comment (helgridly) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/LeoAuthProviderHelper.scala]: Just so I understand: the `TimeoutException` is turned into an `AuthProviderException` in the handler, right?

Or is the `TimeoutException` a subclass of `WorkbenchException` which is handled at the route-level exception handler?

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/ServiceAccountProviderHelper.scala [hunk 7]
```diff
@@ -31,7 +35,11 @@ object ServiceAccountProviderHelper {
   }
 }
 
-class ServiceAccountProviderHelper(wrappedServiceAccountProvider: ServiceAccountProvider, config: Config) extends ServiceAccountProvider(config) with LazyLogging {
+class ServiceAccountProviderHelper(wrappedServiceAccountProvider: ServiceAccountProvider, config: Config)(implicit system: ActorSystem)
+  extends ServiceAccountProvider(config) with FutureSupport with LazyLogging {
+
+  private lazy val providerTimeout = config.getAs[FiniteDuration]("providerTimeout").getOrElse(30 seconds)
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/ServiceAccountProviderHelper.scala]: drive-by nitpick: if we put these default values in `reference.conf` we shouldn't need `getOrElse` everywhere

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/LeoAuthProviderHelper.scala (hunk 5 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/LeoAuthProviderHelper.scala]: The former. So the `TimeoutException` and its message are thrown away and converted to an `AuthProviderException` here (which is handled upstream).

Just thinking, I could tweak the message here so users would see "provider timed out" instead of "provider errored" in this case.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/ServiceAccountProviderHelper.scala (hunk 7 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/ServiceAccountProviderHelper.scala]: True, this is double-defaulted. Will remove the default here.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/sam/SamAuthProvider.scala [hunk 8]
```diff
@@ -1,23 +1,55 @@
...
+    .expireAfterWrite(notebookAuthCacheExpiryTime.toSeconds, TimeUnit.SECONDS)
+    .maximumSize(notebookAuthCacheMaxSize)
+    .build(
+      new CacheLoader[SamCacheKey, Future[Boolean]] {
+        def load(key: SamCacheKey) = {
+          key match {
+            case NotebookAuthCacheKey(userEmail, action, googleProject, clusterName, executionContext) =>
+              hasNotebookClusterPermissionInternal(userEmail, action, googleProject, clusterName)(executionContext)
+            case CanSeeAllClustersInProjectCacheKey(userEmail, googleProject, executionContext) =>
+              canSeeAllClustersInProjectInternal(userEmail, googleProject)(executionContext)
```
Comment (ansingh7115) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/sam/SamAuthProvider.scala]: Cool - I like this way of separating out between the caches - I was imagining 2 separate caches, but this is better i think

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/LeoAuthProviderHelper.scala [hunk 9]
```diff
@@ -1,28 +1,32 @@
...
 
-import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.duration._
+import scala.concurrent.{ExecutionContext, Future, TimeoutException}
 import scala.util.control.NonFatal
 
-case class AuthProviderException(authProviderClassName: String)
-  extends LeoException(s"Call to $authProviderClassName auth provider failed", StatusCodes.InternalServerError)
+case class AuthProviderException(authProviderClassName: String, isTimeout: Boolean = false)
+  extends LeoException(s"Call to $authProviderClassName auth provider ${if (isTimeout) "timed out" else "failed"}", StatusCodes.InternalServerError)
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/LeoAuthProviderHelper.scala]: nice

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/auth/sam/SamAuthProviderSpec.scala [hunk 10]
```diff
@@ -131,4 +134,66 @@ class SamAuthProviderSpec extends TestKit(ActorSystem("leonardotest")) with Free
...
+    // populate backing samClient
+    samAuthProvider.samClient.notebookClusters += (project, name1, userInfo.userEmail) -> Set("sync")
+
+    // call provider method
+    samAuthProvider.hasNotebookClusterPermission(userInfo.userEmail, SyncDataToCluster, project, name1).futureValue shouldBe true
+
+    // cache should contain 1 entry
+    samAuthProvider.notebookAuthCache.size shouldBe 1
+    val key = NotebookAuthCacheKey(userInfo.userEmail, SyncDataToCluster, project, name1, implicitly[ExecutionContext])
+    samAuthProvider.notebookAuthCache.asMap.containsKey(key) shouldBe true
```
Comment (ansingh7115) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/auth/sam/SamAuthProviderSpec.scala]: Why is it necessary to do `asMap` if `samAuthProvider.notebookAuthCache` also has a `get` method?

Comment (ansingh7115) [reply src/test/scala/org/broadinstitute/dsde/workbench/leonardo/auth/sam/SamAuthProviderSpec.scala]: Oh - is it so that we don't cause the cache to refresh potentially?

Comment (rtitle) [reply src/test/scala/org/broadinstitute/dsde/workbench/leonardo/auth/sam/SamAuthProviderSpec.scala]: Right - `cache.get(key)` will call the cache loader if the key is not present.

`cache.asMap().get(key)` won't call the cache loader (you are just operating on the backing `Map`)

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/auth/sam/SamAuthProviderSpec.scala [hunk 11]
```diff
@@ -131,4 +134,66 @@ class SamAuthProviderSpec extends TestKit(ActorSystem("leonardotest")) with Free
...
+
+    // cache should be empty
+    samAuthProvider.notebookAuthCache.size shouldBe 0
+
+    // populate backing samClient
+    samAuthProvider.samClient.billingProjects += (project, userInfo.userEmail) -> Set("list_notebook_cluster")
+
+    // call provider method
+    samAuthProvider.canSeeAllClustersInProject(userInfo.userEmail, project).futureValue shouldBe true
+    samAuthProvider.canSeeAllClustersInProject(WorkbenchEmail("somecreep@example.com"), project).futureValue shouldBe false
```
Comment (ansingh7115) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/auth/sam/SamAuthProviderSpec.scala]: Could potentially put repeated things like `WorkbenchEmail("somecreep@example.com")`

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/sam/SamAuthProvider.scala (hunk 8 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/sam/SamAuthProvider.scala]: Yeah it seemed to make sense because we're caching similar things (sam auth results) which have the same type (Boolean).

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/auth/sam/SamAuthProviderSpec.scala (hunk 11 above)
Comment (ansingh7115) [reply src/test/scala/org/broadinstitute/dsde/workbench/leonardo/auth/sam/SamAuthProviderSpec.scala]: That's definitely something I should have done when I wrote the other tests in this file :)
